Use managed signer for bundles

jtschuster opened the pull request:
Updates the single-file bundler to use the managed Mach-O signer. Adds the SymbolTable load command to the MachObjectFile class and adds a method to extend the symbol table to include the bundled data. Adds validation when creating a MachObjectFile from a file on disk.

~~Now that we don't use codesign for signing, it looks like the executables are sometimes getting killed by antivirus when tests try to run them immediately after they are modified, leading to flaky tests (locally at least). I'm looking into either retrying or running `codesign -v` before running the apps on Mac.~~
The issue wasn't AV, it was due to an issue with signature caching on Mac. The way to work around this is to create a new file with a new inode when modifying and resigning a file on disk. Basic host signing didn't run into this issue because it modifies and signs the executable in the same step, so no re-signing happened. For singlefile bundles, we previously would modify the apphost, close the file, then in a separate step, bundle the files and re-sign it. After this PR we create a new file every time we need to re-sign the host, then replace the old host with the new one.

Fixes https://github.com/dotnet/runtime/issues/110055

jtschuster committed:
Use managed signer for bundles

jtschuster committed:
Fix tests and write out symtab

- Write out the updated symtab command when writing Mach-O file
- Pass macosCodesign through to test CreateAppHost methods
- Remove redundant `codesign` checks
- Warn when bundler is told to sign the bundle for a non-macos target

jtschuster committed:
Fix test failures

- Allow macosCodesign to be true for non-mac bundles
- Allow some padding between string table and signature

jtschuster committed:
Flip comparison

jtschuster committed:
Add comment to TryAdjustHeadersForBundle, and throw if it returns false

jtschuster committed:
Don't memory-map file if it's not a Mach-O, remove unused file.

jtschuster committed:
Fix typo, restructure IsSigned check.

elinor-fung reviewed (commented):
> Now that we don't use codesign for signing, it looks like the executables are sometimes getting killed by antivirus when tests try to run them immediately after they are modified, leading to flaky tests (locally at least). I'm looking into either retrying or running codesign -v before running the apps on Mac.

What was the end result here? Could a product scenario hit the same thing with running immediately after publish with signing?

jtschuster commented on src/installer/managed/Microsoft.NET.HostModel/MachO/MachObjectFile.cs:
```diff
@@ -296,13 +329,46 @@ private static long ReadCommands(
+            // Signature blob should be right after the symbol table except for a few bytes of padding
+            uint symtabEnd = symtabLC.Command.GetStringTableOffset(header) + symtabLC.Command.GetStringTableSize(header);
+            uint signStart = codeSignatureLC.Command.GetDataOffset(header);
+            if (symtabEnd > signStart || signStart - symtabEnd > 32)
```
Anecdotally, it looks like 16 byte alignment, but I haven't bee able to find documentation for that.

jtschuster commented on src/installer/managed/Microsoft.NET.HostModel/Bundle/Bundler.cs:
```diff
@@ -345,18 +345,7 @@ public string GenerateBundle(IReadOnlyList<FileSpec> fileSpecs)
-                    throw new InvalidOperationException($"Failed to codesign '{bundlePath}': {stdErr}");
-                }
-            }
+            HostWriter.SetAsBundle(bundlePath, headerOffset, _macosCodesign);
```
Good catch, we can move it out

jtschuster committed:
PR Feedback: Fomatting and comments

jtschuster committed:
PR feedback:

- Use the same memory-mapped file instance for placeholder replacement
  and signing
- formatting changes

jtschuster commented:
> What was the end result here?

It looks like the issue is probably related to modifying the exe in place: https://developer.apple.com/documentation/security/updating-mac-software#Update-Files-that-Include-Signed-Code

I'll update the code to use CopyOnWrite memory-mapped files, write a new file, then replace the old one.

jtschuster committed:
Create new file for macos signing, update LINKEDIT.VMSize to fit signature

jtschuster commented:
> > > What was the end result here?
> > 
> > 
> > It looks like the issue is probably related to modifying the exe in place: https://developer.apple.com/documentation/security/updating-mac-software#Update-Files-that-Include-Signed-Code
> 
> Ah, does it cache once the file is created? I had thought it would only cache when it is first run.
> 
> > I'll update the code to use CopyOnWrite memory-mapped files, write a new file, then replace the old one.
> 
> I guess doing the inode check with `codesign` shows it replaces with an entirely new file? Matching that seems good.

It was still broken after changing the behavior to create a new inode. I had forgotten to update the VMSize field of the __LINKEDIT segment when creating the signature. The file size was larger than the VMSize, and that was causing sporadic crashes, but I only noticed it in the larger bundles with larger signatures. I left in the "copy to a temp file, edit, then replace" behavior anyway to make sure we match `codesign`'s behavior.

jtschuster committed:
Merge branch 'main' into ManagedSignBundles

jtschuster committed:
Merge branch 'main' into ManagedSignBundles

elinor-fung commented on src/installer/tests/Microsoft.NET.HostModel.Tests/MachObjectSigning/SigningTests.cs:
```diff
@@ -24,18 +24,26 @@ namespace Microsoft.NET.HostModel.MachO.CodeSign.Tests
+            using (var memoryMappedFile = MemoryMappedFile.CreateFromFile(appHostSourceStream, null, 0, MemoryMappedFileAccess.Read, HandleInheritability.None, true))
+            using (var managedSignedAccessor = memoryMappedFile.CreateViewAccessor(0, 0, MemoryMappedFileAccess.CopyOnWrite))
+            {
+                if (!MachObjectFile.Create(managedSignedAccessor).HasSignature) {
+                    return false;
+                }
+            }
+            if (Codesign.IsAvailable
+                && Codesign.Run("--verify", filePath).ExitCode != 0) {
+                    return false;
+            }
```
nit: opening brace on new line.

elinor-fung commented on src/installer/tests/TestUtils/CommandRetry.cs:
```diff
@@ -0,0 +1,88 @@
+
+namespace Microsoft.DotNet.Cli.Build.Framework
+{
+    public class RetriedCommand
```
Not used / needed?

elinor-fung commented on src/installer/managed/Microsoft.NET.HostModel/MachO/MachObjectFile.cs:
```diff
@@ -232,7 +271,13 @@ static bool LinkEditSegmentsAreEquivalent((Segment64LoadCommand Command, long Fi
     public static long GetSignatureSizeEstimate(uint fileSize, string identifier)
     {
-        return CodeSignature.GetCodeSignatureSize(fileSize, identifier);
+        return CodeSignature.GetCodeSignatureSize(fileSize, identifier) + (AlignUp(fileSize, 0x10) - fileSize);
```
Add a const for 0x10?

elinor-fung commented on src/installer/managed/Microsoft.NET.HostModel/MachO/MachObjectFile.cs:
```diff
@@ -232,7 +271,13 @@ static bool LinkEditSegmentsAreEquivalent((Segment64LoadCommand Command, long Fi
+        return CodeSignature.GetCodeSignatureSize(fileSize, identifier) + (AlignUp(fileSize, 0x10) - fileSize);
+    }
+
+    public static long GetSignedFileSizeEstimate(string filePath)
```
What's using this?

jtschuster replied:
Nothing is using this - good catch.

jtschuster committed:
PR feedback

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into ManagedSignBundles

jtschuster committed:
Remove old file that's not needed

jtschuster committed:
Merge branch 'ManagedSignBundles' of https://github.com/jtschuster/runtime into ManagedSignBundles

jtschuster committed:
Merge branch 'main' of https://github.com/dotnet/runtime into ManagedSignBundles

elinor-fung commented on src/installer/managed/Microsoft.NET.HostModel/MachO/MachObjectFile.cs:
```diff
@@ -232,7 +271,8 @@ static bool LinkEditSegmentsAreEquivalent((Segment64LoadCommand Command, long Fi
     public static long GetSignatureSizeEstimate(uint fileSize, string identifier)
     {
-        return CodeSignature.GetCodeSignatureSize(fileSize, identifier);
+        const int CodeSignatureSizeAlignment = 0x10;
```
Use the same constant in `GetSignatureStart`?

jtschuster committed:
Use constant for alignments

jtschuster committed:
Merge branch 'main' into ManagedSignBundles
